### DISCUSSION TITLE
Cancel previous build/cancel deployment when action is cancelled

### DISCUSSION THREAD

**Discussion by shouze** [BetaHuhn/deploy-to-vercel-action#94] [Ideas]
When you push/update some branch more quickly than vercel is able to build them... would be nice to cancel previous builds instead of letting them queued until completion.

I don't know how we could proceed at the moment, I just want to open the discussion at this point to find a way to achieve that.

**Comment by BetaHuhn** [comment]
Yeah seems reasonable.

We could use the [Vercel API](https://vercel.com/docs/rest-api#endpoints/deployments/cancel-a-deployment) to cancel a deployment.

On each run we would have to check if there’s a previous unfinished deployment for that branch and then cancel it before creating a new one.

PR welcome!

**Reply by shouze** [reply]
In fact [github actions looks to have a very useful concurrency parameter for jobs](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#concurrency), so it manages queued concurrent jobs & cancellation of them. The only thing to ensure would be that when a job is canceled a signal is effectively sent to the job and that vercel cli can handle this signal to cancel in progress build.

**Reply by BetaHuhn** [reply]
[Looks like](https://github.community/t/graceful-job-termination/121103/3) GitHub just sends a SIGINT to the Node process and handles it quite gracefully, so we could do it like this:

```js
process.on('SIGINT', function() {
  // Cancel build here, then exit
  process.exit();
})
```

**Reply by shouze** [reply]
Bad news is it looks like I have this message logged
![image](https://user-images.githubusercontent.com/54712/135809028-d73309e5-d791-48c7-b40d-392313597f6e.png)

but the related build in vercel console is not canceled... so I guess we probably have to cancel it by invoking another vercel cli command. I'm not familiar of vercel cli possibilities, are you? If yes it could speedup time to PR, otherwise I'll take a look.

**Reply by shouze** [reply]
I didn't find a way in vercel cli documentation to cancel a build. Did you?

But we can [through api](https://vercel.com/docs/rest-api#endpoints/deployments/cancel-a-deployment)

This imply to have the build id, so does it mean that it also imply to [manually create the deployment through api](https://vercel.com/docs/rest-api#endpoints/deployments/create-a-new-deployment) instead of using vercel cli, then poll it until completion (success/failure)?

**Reply by BetaHuhn** [reply]
Yes, it’s only possible with the API, as I mentioned [above](https://github.com/BetaHuhn/deploy-to-vercel-action/discussions/94#discussioncomment-1422964).

> This imply to have the build id, so does it mean that it also imply to manually create the deployment through api instead of using vercel cli, then poll it until completion (success/failure)?

No, we are already retrieving the ID from the API manually after deploying via the CLI, see here: https://github.com/BetaHuhn/deploy-to-vercel-action/blob/98dbe809764ce8219ce6ba341000e912bc328d2c/src/index.js#L104

We can just reuse that to get the ID to cancel the deployment. The only change we need to do is to store it globally so we can have access to it from the `process.on()` handler.

**Reply by BetaHuhn** [reply]
I‘d also suggest that we introduce a new action option which enables/disables this behavior. Some people might not want to cancel the Vercel Deployment when the Action run is canceled. Maybe something like `CANCEL_DEPLOYMENT_WITH_RUN`, or something else if you have a better suggestion :)

**Reply by shouze** [reply]
Agreed.
To keep consistency with Github Actions job concurrency naming I suggest `CANCEL_IN_PROGRESS_DEPLOYMENT` for the option name. WDYT?

**Reply by BetaHuhn** [reply]
Yes makes sense. Lets have it default to false to start of with so we don’t have to release this as a breaking change.

Do you feel comfortable creating a PR for this?

**Reply by shouze** [reply]
Will try yes, the biggest difficulty to me is that we cannot use vercel cli anymore to create the deployment, because from what I've seen the vercel cli doesn't return the build id unless the build has finished or failed.
![image](https://user-images.githubusercontent.com/54712/135822411-eb2f8ab0-67cd-49fd-84e6-264a5ce5de3d.png)


 The best option would probably be to replace [vercel](https://github.com/BetaHuhn/deploy-to-vercel-action/blob/98dbe809764ce8219ce6ba341000e912bc328d2c/package.json#L30) by [@vercel/client](https://github.com/vercel/vercel/tree/main/packages/client), WDYT?

**Reply by shouze** [reply]
Good news it that we will be able to print out progress dots while we poll the progress of the deployment ^^
https://github.com/vercel/vercel/blob/main/packages/client/src/check-deployment-status.ts

**Reply by BetaHuhn** [reply]
For the time being I don’t want to refactor the whole action to use `@vercel/client` instead of the cli as we can achieve the same result without much modification.

You can just copy and modify the `getDeployment` function which uses the API:

https://github.com/BetaHuhn/deploy-to-vercel-action/blob/98dbe809764ce8219ce6ba341000e912bc328d2c/src/vercel.js#L78-L89

You only need to replace the endpoint with the one from the [Vercel API docs](https://vercel.com/docs/rest-api#endpoints/deployments/cancel-a-deployment).

**Reply by shouze** [reply]
Ok... but getDeployment needs deploymentUrl... and I can't obtain it unless vercel cli has finished the deployment, right? So to me we're stuck if we don't replace vercel cli by `@vercel/client`. Anyway if you have any idea to hack that and reduce efforts I'm in, as I would pefer minor changes with big impact.

**Reply by BetaHuhn** [reply]
Ah I see your point now.

We could work around this by using the [List Deployments](https://vercel.com/docs/rest-api#endpoints/deployments/list-deployments) endpoint of the Vercel API to get the deployment without knowing it’s id by matching the `meta-[key]` query parameter against one of the ones we provide (like the git ref or sha): 

https://github.com/BetaHuhn/deploy-to-vercel-action/blob/98dbe809764ce8219ce6ba341000e912bc328d2c/src/vercel.js#L36-L46

Will have to check if that would actually work though.

How would you imagine it working with `@vercel/client`? Does it return the deployment id before it has finished?

**Reply by shouze** [reply]
Looks like `@vercel/client` returns the deployment id before it has finished yes, I need to double check... but if it's the case would be easier to replace `vercel` cli by `@vercel/client` at the end.

**Reply by shouze** [reply]
![image](https://user-images.githubusercontent.com/54712/135898105-bce8006b-210e-45a8-a4af-27b11ba687ac.png)

**Reply by shouze** [reply]
By listening `created` event we will have the id so this would be easy to cancel the deployment on `SIGINT`.

But cancel is not included into `@vercel/client` at the moment so for this one it will require to directly call the [Vercel API](https://vercel.com/docs/rest-api#endpoints/deployments/cancel-a-deployment) as you previously mentioned.

**Reply by BetaHuhn** [reply]
Okay, thanks for checking.

I hacked around with the CLI a bit and I think we might be able to get the deployment URL or even the id during the deployment creation using the CLI. The CLI writes the deployment URL to [stdout](https://vercel.com/docs/cli#commands/overview/standard-output-usage) as soon as the deployment is created, so we just need to wait for it to write it, store it somewhere and then wait for the command to finish like we do now.

If I can get it to work with the above method, I think we should stick to the CLI it as it doesn't justify rewriting the action to use `@vercel/client` (as we already agreed). I will try both your option with `@vercel/client` and the CLI tomorrow and see what works best.

PS: Thanks for all your help!

**Reply by BetaHuhn** [reply]
I ended up getting it to work without `@vercel/client` or any weird hacks with the CLI. The action is now attaching the [run id](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context) (a unique number for each action run within a repository) as metadata to the Vercel deployment and using that to later find the deployment [via the API](https://vercel.com/docs/rest-api#endpoints/deployments/list-deployments) in the `SIGINT` handler. This way it can find the deployment without knowing the deployment url or id.

I choose this method because it requires the least amount of refactoring and because it seems like a stable solution. There's a working version of this available on the [feat/cancel-build](https://github.com/BetaHuhn/deploy-to-vercel-action/tree/feat/cancel-build) branch which you can try out.

Some things to consider/think about:

- The `SIGINT` handler gets called every time the process of the action is being terminated/killed, I don't know if there's another case besides the action being cancelled when this happens
- I think it might be worth to enable this new behavior by default as it is somewhat unexpected for the build to go on even when you cancel the action run. I don't think this qualifies as a breaking change and would suggest to release this as a `minor` change
- I have only tested the implementation by cancelling the action run via the GitHub UI, not by pushing another commit while another deployment is still being built (i.e. your original issue)
- The GitHub Deployment is now also being set to `inactive` when the action gets cancelled

Let me know if there's something I missed/any thoughts you have about this.

**Reply by shouze** [reply]
Good move!

However, here's what I get when I push over a PR with an already in progress vercel build:
![image](https://user-images.githubusercontent.com/54712/136281945-f2c56457-9cad-45ed-9463-570640469e01.png)


Looks like this is probably because this call returns a 404:
https://github.com/BetaHuhn/deploy-to-vercel-action/commit/d4ff77d17f2b3dc812b944240babd142dda8c80d#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556R186

**Reply by shouze** [reply]
And yes, should be a good idea to enable this new behavior by default :zap:

**Reply by shouze** [reply]
I even tried to inject by myself env var `RUN_ID` with `${{ github.ref }}` value but with no luck :sob: 

In fact, when pushing over an existing PR, the way `RUN_ID` is built makes it unique anyway as [it is suffixed by `-${ Date.now() }`](https://github.com/BetaHuhn/deploy-to-vercel-action/commit/33d136e1283a4c6679c2724a6ef9f22077f1ee71#diff-3d2b59189eeedc2d428ddd632e97658fe310f587f7cb63b01f9b98ffc11c0197R13990-R14000
).

So `getDeploymentByRunId` will always return a 404, except when you cancel manually a running Github workflow, it will this time cancel correctly the Vercel build. I suggest that if we want to inject `RUN_ID`, you shouldn't suffix it at all, WDYT?

**Reply by shouze** [reply]
**Edit**: for an unknown reason the vercel build is not canceled either when I cancel github workflow manually. Maybe because I was still injecting my custom `RUN_ID`?

**Reply by shouze** [reply]
**Edit2**: Even if I revert custom `RUN_ID`, when I manually cancel the Github Workflow I still have the very same 404 error on cleanup
![image](https://user-images.githubusercontent.com/54712/136287597-de5c07b5-7786-4667-a2cc-f25dbeaa7be6.png)

**Reply by BetaHuhn** [reply]
I chose to append `RUN_ID` with `Date.now()` as `RUN_ID` will be the same if you re-run a Workflow. By appending the time, it will be unique every time it runs. The calculated value will be the same during the run of the workflow.

It worked when I tested it by cancelling the workflow run manually, I don't know why you are getting a 404. Maybe it didn't have enough time to create the deployment when you cancelled the run?

**Reply by shouze** [reply]
Yes undestood regarding `RUN_ID`, but when you push over and over on the same branch, it will generate distincts `RUN_ID` values and in fact this is not what we want. For PR my guess is that we want a `RUN_ID` with is the PR uniquement identifier or something like that.

About the 404, yes very weird, it had time to created Vercel builds I've double checked by having Vercel console opened aside :wink:

**Reply by BetaHuhn** [reply]
Ah yeah makes sense. I wasn't thinking in regards to pushing over and over to the same branch, only about cancelling a single run.

We need to replace `RUN_ID` with an identifier for a given branch. We could use the [git ref](https://github.com/BetaHuhn/deploy-to-vercel-action/blob/d4ff77d17f2b3dc812b944240babd142dda8c80d/src/config.js#L121) for that.
